clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -110,78 +112,107 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
+                        remainingTimeToBlockNs -= timeNs;
+                        // check if we can satisfy this request from the free list,
+                        // otherwise allocate memory
+                        if (accumulated == 0 && size == this.poolableSize && !this.free.isEmpty()) {
+                            // just grab a buffer from the free list
+                            buffer = this.free.pollFirst();
+                            accumulated = size;
+                            restoreAvailableMemoryOnFailure = false;
+                        } else {
+                            // we'll need to allocate memory, but we may only get
+                            // part of what we need on this iteration
+                            freeUp(size - accumulated);
+                            int got = (int) Math.min(size - accumulated, this.availableMemory);
+                            this.availableMemory -= got;
+                            accumulated += got;
+                        }
                     }
 
-                    remainingTimeToBlockNs -= timeNs;
-                    // check if we can satisfy this request from the free list,
-                    // otherwise allocate memory
-                    if (accumulated == 0 && size == this.poolableSize && !this.free.isEmpty()) {
-                        // just grab a buffer from the free list
-                        buffer = this.free.pollFirst();
-                        accumulated = size;
-                    } else {
-                        // we'll need to allocate memory, but we may only get
-                        // part of what we need on this iteration
-                        freeUp(size - accumulated);
-                        int got = (int) Math.min(size - accumulated, this.availableMemory);
-                        this.availableMemory -= got;
-                        accumulated += got;
-                    }
-                }
-
-                // remove the condition for this thread to let the next thread
-                // in line start getting memory
-                Condition removed = this.waiters.removeFirst();
-                if (removed != moreMemory)
-                    throw new IllegalStateException("Wrong condition: this shouldn't happen.");
+                    // remove the condition for this thread to let the next thread
+                    // in line start getting memory
+                    Condition removed = this.waiters.removeFirst();
+                    if (removed != moreMemory)
+                        throw new IllegalStateException("Wrong condition: this shouldn't happen.");
 
+                    if (buffer == null)
+                        buffer = allocateByteBuffer(size);
+                    restoreAvailableMemoryOnFailure = false;
+                    //unlock happens in top-level, enclosing finally
+                    return buffer;
+                } finally {
```

Comment:
I think there is actually a reasonable way to merge the two `finally` into one. The idea is that the code in the first finally block doesn't throw exception and is no-op if `this.availableMemory + freeListSize >= size`. This is no-op if you change it to the following:

```
if (hasError)
  this.availableMemory += accumulated;
if (moreMemory != null)
  this.waiters.remove(moreMemory);
```